### ISSUE TITLE
Proposal: Add 'is not null' operator

### ISSUE THREAD

**Issue by lellid** [dotnet/csharplang#1944]
Currently, we already have something like
```
if(a is null)
  …
```
The opposite case, the question if something is *not* null, is (asymmectrically) much harder to formulate. It would be nice to have something like
```
if(b is not null)
```
or ideally
```
if(b is not null bnn)
```
so that `bnn` is then guaranteed to be not null in a thread safe way, would make life much easier.

**Comment by CyrusNajmabadi** [comment]
`if(b is not null bnn)` will be able to be written as `if (b is {} bnn)` once 'property patterns' go in.

**Comment by jaredpar** [comment]
Pretty sure `is not null` falls out of pattern matching work. @gafter

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/27
https://github.com/dotnet/csharplang/issues/157
https://github.com/dotnet/csharplang/issues/246
https://github.com/dotnet/csharplang/issues/568
https://github.com/dotnet/csharplang/issues/1350
https://github.com/dotnet/csharplang/issues/1913

**Comment by markusschaber** [comment]
@willard720 `if (!b)`already has defined semantics if b is `bool` or `bool?`, and AFAICS every type which has an implicit conversion to bool.

**Comment by gafter** [comment]
> Pretty sure `is not null` falls out of pattern matching work.

Not until we do https://github.com/dotnet/csharplang/issues/1350 which is not currently expected to hit C# 8.0.

**Comment by willard720** [comment]
@markusschaber True, but most people don't do an implicit conversion to bool, and I can't imagine what else would make sense for it. If someone did the code `if (employee)` instead of `if (employee.IsPunchedIn)` then they're using bad practices, so `if (employee)` will almost always mean `not null`.

**Comment by markusschaber** [comment]
@willard720 It just might be confusing, and it might not be easy to resolve the conflicts with existing semantics ambigously.

e. G. for `bool?`, will it mean `not null` or `equal to true`?

Other languages have rather intuitive concepts of "truth value" (like [Python,](https://docs.python.org/2.4/lib/truth.html) where thinks like 0, 0.0, `None` and empty collections or strings evaluate to false, while their non-emtpy counterparts evaluate to true). It feels natural, and I actually like it, but I'm not sure whether it's possible to introduce such a concept in C# now, without breaking existing code. 

And I doubt that being `null` alone is the best criteria for "truth values", I'd rather see some concept like in Python.

However, I'm afraid we're off-topic for this issue. :-(

**Comment by MkazemAkhgary** [comment]
```c#
if(a is not null)
if(a != null)
```

This ancient syntax is still better. for programmers its easy to catch. 

And what is `bnn` for? `a` is already garaunteed to be not null inside that block. I dont get the thread safety argument.

**Comment by CyrusNajmabadi** [comment]
> And what is bnn for? a is already guaranteed to be not null inside that block. I dont get your thread safety argument.

If 'a' is not a local/parameter, there is no guarantee that it will be non-null, even after checking it against null.  Specifically, another thread could end up changing it's value.  Patterns, allow you to perform tests, then place the results in a local when successful, so you have total certainty about the state of the variable if the pattern succeeds in matching.

**Comment by MkazemAkhgary** [comment]
@CyrusNajmabadi Im illuminated! 🌝

**Comment by markusschaber** [comment]
@RichardD2 I did not even know about those operators....
I'm doing C# for 10 years now, and I still learn something new every day... :-)

**Comment by RichardD2** [comment]
Surely `if (b is not null) ...`  can already be written in C# 7 as `if (b is object) ...`? A `null` reference will only match the `null` or `var` patterns.

The compiler even seems to correctly handle value types _(generates a CS0183 warning, and omits the test from IL)_ and nullable value types _(replaces the test with a call to `b.HasValue`)_.

Of course, it wouldn't work for `b is not null bnn`, since you'd presumably want `bnn` to have the same compile-time type as `b`, whereas with `b is object bnn`, it would be an `object`.

**Comment by YairHalberstadt** [comment]
Closing as `not` patterns have been implemented for C# 9.0 as part of #1350

Whilst `if(b is not null bnn)` is not currently allowed, you can do `if(b is {} bnn)` with the equivalent meaning, and if the championed https://github.com/dotnet/csharplang/issues/3369 is implemented `is not null bnn` would have a different meaning.
